GG/src/RichText/ImageBlock.cpp

```diff
@@ -0,0 +1,150 @@
+        try {
+            boost::shared_ptr<Texture> texture = GetTextureManager().GetTexture (image_path);
+            m_graphic = new StaticGraphic (texture, GRAPHIC_PROPSCALE | GRAPHIC_SHRINKFIT | GRAPHIC_CENTER);
+            AttachChild (m_graphic);
+        } catch (GG::Texture::BadFile& errBadFile) {
+            // No can do inside GiGi.
+        }
+    }
+
+    Pt ImageBlock::SetMaxWidth (X width) {
+
+        if (m_graphic != 0) {
+            // Give the graphic the set width and give it liberty with the height.
+            m_graphic->Resize (Pt (width, Y (INT_MAX)));
+
+            // Get the actual space the graphic decided to use.
+            Rect rect = m_graphic->RenderedArea();
+            Pt size = rect.LowerRight() - rect.UpperLeft();
+
+            // Take up the full width to center the image.
+            size.x = width;
+
+            // Don't take the extra vertical space.
+            m_graphic->Resize (size);
+
+            // Update our size to match the graphic we are wrapping.
+            Resize (size);
+
+            // Return the size we decided to be.
+            return size;
+        } else {
+            // We don't have an image. Take a quarter of width to show an X.
+            Pt size (width, Y (Value (width) / 4));
+            Resize (size);
+            return size;
+        }
+    }
+
+    void ImageBlock::Render() {
+        if (m_graphic == 0) {
+            // Error: no image. Draw red x.
+            Pt ul = UpperLeft();
+            Pt lr = LowerRight();
+            Pt size = lr - ul;
+            ul.x += size.x / 2 - X (Value (size.y)) / 2;
+            lr.x -= size.x / 2 - X (Value (size.y)) / 2;
+            FlatX (ul, lr, CLR_RED);
+        }
+    }
+
+// A factory for creating image blocks from tags.
+    class ImageBlockFactory: public RichText::IBlockControlFactory {
+        public:
+            ImageBlockFactory() :
+                m_rootPath (".")
+            {}
+
+            //! Create a Text block from a plain text tag.
+            virtual BlockControl* CreateFromTag (const std::string& tag,
+                                                 const RichText::TAG_PARAMS& params,
+                                                 const std::string& content,
+                                                 const boost::shared_ptr<Font>& font,
+                                                 Clr color,
+                                                 Flags<TextFormat> format
+                                                ) {
+                // Get the path from the parameters.
+                std::string path = ExtractPath (params);
+
+                // Create a new image block, basing the path on the root path.
+                return new ImageBlock ( (m_rootPath / path).string() , X0, Y0, X1, Flags<WndFlag>());
+            }
+
+            // Sets the root of image search path.
+            void SetRootPath (const std::string& path) {
+                m_rootPath = path;
```

Comment:
double check that setting path with a string in this way is safe. in other places, utf-8 conversions are done